### ISSUE TITLE
Key completion on class

### ISSUE THREAD

**Issue by DavidMStraub** [ipython/ipython#10903]
The dictionary key completion feature and the possibility to define custom completion methods via `_ipython_key_completions_` are great, but I have a use case where it doesn't work: I have a class that keeps track of its instances, that all have (string) names, via a dictionary attribute. By defining `__getitem__` on the class's metaclass, I can access instances via their name like so:

```python
my_instance = NamedInstanceClass['name of my instance']
```

[See here for the definition](https://github.com/flav-io/flavio/blob/master/flavio/classes.py#L18-L68). Now, the reason why key completion doesn't work in this case are [these two lines](https://github.com/ipython/ipython/blob/master/IPython/utils/dir2.py#L63-L64) in `dir2` that exclude completion for classes. Indeed if I monkey-patch that method like

```python
def my_getattr(obj, method):
    return getattr(obj, method, None)
IPython.core.completer.get_real_method = my_getattr
```

then key completion works perfectly by defining a class method

```python
@classmethod
def _ipython_key_completions_(cls):
  return cls.instances.keys()
```

on the `NamedInstanceClass`.

If you agree that this is something that should/can be fixed, I would be happy to submit a PR.

System info:
```python
{'commit_hash': 'ca5443062',
 'commit_source': 'installation',
 'default_encoding': 'UTF-8',
 'ipython_path': '/usr/local/lib/python3.6/dist-packages/IPython',
 'ipython_version': '6.2.1',
 'os_name': 'posix',
 'platform': 'Linux-4.13.0-17-generic-x86_64-with-Ubuntu-17.10-artful',
 'sys_executable': '/usr/bin/python3',
 'sys_platform': 'linux',
 'sys_version': '3.6.3 (default, Oct  3 2017, 21:45:48) \n[GCC 7.2.0]'}
```

**Comment by takluyver** [comment]
I did a bit of git-blaming: the check was added in #7006, and it did fix some problems at the time.

I think the nature of Python can make it tricky to tell whether we're getting a class method or an unbound instance method, especially where different types of classes are involved (e.g. classes from compiled modules).

@minrk do you remember more about your thinking at the time you wrote the check?

**Comment by minrk** [comment]
I suspect the check I added was naïve, not handling the case of classes properly. But #7006 is about display methods, not completer methods, though the logic is probably similar and with similar reasoning behind it.

If I recall correctly, the issues are related to compiled methods on objects, which are harder to identify whether methods are static, class, or instance methods. This was the case that the more complicated checks tried and failed to identify methods reliably across Python versions and for compiled and non-compiled objects. So the fix was to simply punt: explicitly exclude classes themselves from magic-method pretty representation lookup. That way we don't run into issues where `Class._repr_foo_` is defined, but we can't tell if it should be used for representing the *class* or *instances* of the class.

We can probably whitelist this particular case: obj is a type and the method is clearly identified as a class method (`isinstance(meth, types.MethodType)` *should* work, I think), they can be used on the class, otherwise assume they will be instance methods. This will be unavailable to most compiled classes (okay, same as now), but avoid touching the instance case.

**Comment by minrk** [comment]
Got it, that makes sense.

**Comment by ckingdev** [comment]
I'm just going out on a limb here, I am not familiar with the internals really, but is this possibly related to jedi being very slow for some people who have large objects in memory? 

Looking back, I've often had very large dictionaries/counters in memory when I've had trouble. If jedi was trying to go through all the possible keys, that would explain why I've had sometimes up to a minute of not being able to type after hitting tab. I thought it was the GC for a while but it's not a problem if I don't try to use tab complete. It would also explain why I have not had problems when using a trie that didn't overload __get__().

**Comment by takluyver** [comment]
Fixed by #10905, if I understood correctly.
